odoo/addons/base/models/ir_ui_view.py

```diff
@@ -581,62 +584,84 @@ def _get_filter_xmlid_query(self):
                   WHERE res_id IN %(res_ids)s AND model = 'ir.ui.view' AND module IN %(modules)s
                """
 
-    def get_inheriting_views_arch(self, model):
-        """Retrieves the sets of views that should currently be used in the
-           system in the right order. During the module upgrade phase it
-           may happen that a view is present in the database but the fields it relies on are not
-           fully loaded yet. This method only considers views that belong to modules whose code
-           is already loaded. Custom views defined directly in the database are loaded only
-           after the module initialization phase is completely finished.
-
-           :param str model: model identifier of the inheriting views.
-           :return: list of ir.ui.view
+    def _get_inheriting_views(self):
         """
-        self.ensure_one()
-        self.check_access_rights('read')
-
-        # retrieve all the views transitively inheriting from view_id
-        domain = self._get_inheriting_views_arch_domain(model)
+        Determine the views that inherit from the current recordset, and return
+        them as a recordset, ordered by priority then by id.
+        """
+        domain = self._get_inheriting_views_domain()
         e = expression(domain, self.env['ir.ui.view'])
         from_clause, where_clause, where_params = e.query.get_sql()
-        assert from_clause == '"ir_ui_view"'
-        self.flush(['active'])
-        query = """
+        sub_where_clause = where_clause.replace('ir_ui_view', 'view')
+        assert from_clause == '"ir_ui_view"', f"Unexpected from clause: {from_clause}"
+
+        self._flush_search(domain, fields=['inherit_id', 'priority', 'model', 'mode'], order='id')
+        query = f"""
             WITH RECURSIVE ir_ui_view_inherits AS (
-                SELECT id, inherit_id, priority
+                SELECT id, inherit_id, priority, mode, model
                 FROM ir_ui_view
-                WHERE inherit_id = %s AND {where_clause}
+                WHERE id IN %s AND {where_clause}
             UNION
-                SELECT iuv.id, iuv.inherit_id, iuv.priority
-                FROM ir_ui_view iuv
-                INNER JOIN ir_ui_view_inherits iuvi ON iuvi.id = iuv.inherit_id
-                WHERE {sub_where_clause}
+                SELECT view.id, view.inherit_id, view.priority, view.mode, view.model
+                FROM ir_ui_view view
+                INNER JOIN ir_ui_view_inherits parent ON parent.id = view.inherit_id
+                WHERE coalesce(view.model, '') = coalesce(parent.model, '')
+                      AND view.mode = 'extension'
+                      AND {sub_where_clause}
```

Comment:
same thing
```suggestion
                      AND ({sub_where_clause})
```

Reply:
There are parenthesis already, see expression.py L614

```where_clause: '((FALSE OR "ir_ui_view"."website_id" IS NULL) AND ("ir_ui_view"."active" = %s))'```

Reply:
> There are parenthesis already, see expression.py L614

The braces are an implementation detail that is not guaranteed at all by `get_sql()`, so we *should* include extra braces. It's trivial to obtain a `where_clause` that isn't surrounded by braces via `get_sql()`, whereas there is no problem with having an extra pair of braces.

I understand that it's not your code, but it should have been caught in earlier reviews.

That's safer, defensive programming, and we want the framework to be solid and defensive.

Reply:
done